flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java

```diff
@@ -0,0 +1,830 @@
+            // with the output of this join level, which is a null padded row.
+            processWithNullPadding(depth, input, inputId, currentRows, associations, phase);
+        }
+    }
+
+    // This can simply emit the resulting join row between all n inputs.
+    private void emitJoinedRow(RowData input, RowData[] currentRows) {
+        // TODO Gustavo: check if we don't have to check the multi join condition here
+        emitRow(input.getRowKind(), currentRows);
+    }
+
+    // Problem: we are not using the primary key but rather always the partitioning key
+    private boolean processRecords(
+            int depth,
+            RowData input,
+            int inputId,
+            RowData[] currentRows,
+            int[] associations,
+            JoinPhase phase,
+            boolean isLeftJoin)
+            throws Exception {
+        boolean matched = false; // Tracks if any record at this depth matched the condition
+
+        // This optimization avoids that by skipping state iteration at the input's own level.
+        // For inner joins, the input record itself will be processed at the `inputId`
+        // level in `processInputRecord`. We do not have to process any records in state for
+        // the inputId if it's an inner join, since we do not need to know if we the left
+        // side has associations. We only need to know if the left side has associations if
+        // we are doing a left join and need to know if we need to emit a null padded row.
+        if (isInnerJoin(isLeftJoin) && inputId == depth) {
+            return false;
+        }
+
+        // Calculate the joinKey to retrieve from the state only the records that can potentially
+        // match based on the equi-join conditions.
+        // The joinKey consists of all attributes present in equi-join conditions for the current
+        // level.
+        // We use the left side (currentRows) to calculate its value.
+        RowData joinKey = keyExtractor.getJoinKeyFromCurrentRows(depth, currentRows);
+        Iterable<RowData> records = stateHandlers.get(depth).getRecords(joinKey);
+
+        for (RowData record : records) {
+            currentRows[depth] = record;
+
+            // Shortcircuit: if the join condition fails, this path yields no results
+            // we can go to the next record in the state.
+            if (matchesCondition(depth, currentRows)) {
+                matched = true;
+            } else {
+                continue;
+            }
+
+            // For LEFT joins, association counts are updated for the preceding level (depth - 1)
+            // to correctly track if the left-side row found any matches on this right-side.
+            // This information is crucial for determining if null padding is needed later.
+            if (isLeftJoin) {
+                updateAssociationCount(
+                        depth, associations, shouldIncrementAssociation(phase, input));
```

Comment:
it looks like we do not need to calculate `shouldIncrementAssociation(phase, input)` on each loop iteration and can do it before the loop or did I miss anything?

Reply:
We increment the number of associations for ´current depth - 1` for each record, in case they match. We first check if the join conditions matches for each record inside the loop, if it doesn't, we don't want to increment the number of associations and just continue to the next record.

```
if (matchesCondition(depth, currentRows)) {
                matched = true;
            } else {
                continue;
            }
```